src/scanner.cc

```diff
@@ -151,59 +155,184 @@ struct Scanner {
+        // easier to handle line continuations if comments consume the
+        // newline
+
+        // Semicolons and EOF always end the statement
+        if (lexer->lookahead == ';' || lexer->eof(lexer)) {
+          skip(lexer);
+          lexer->result_symbol = _END_OF_STATEMENT;
+          return true;
         }
 
-        // always check for line cont. if nothing else found
-        lexer->result_symbol = _LINE_CONTINUATION;
+        // If we're in a line continuation, then don't end the statement
+        if (in_line_continuation) return false;
 
-        // Consume '&' at the end of the line
-        if (lexer->lookahead != '&') {
+        // Consume end of line characters, we allow '\n', '\r\n' and
+        // '\r' to cover unix, MSDOS and old style Macintosh.
+        // Handle comments here too, but don't consume them
+        if (lexer->lookahead == '\r') {
+          skip(lexer);
+          if (lexer->lookahead == '\n') {
+              skip(lexer);
+          }
+        } else {
+          if (lexer->lookahead == '\n') {
+            skip(lexer);
+          } else if (lexer->lookahead != '!') {
+            // Not a newline and not a comment, so not an
+            // end-of-statement
             return false;
+          }
         }
+
+        lexer->result_symbol = _END_OF_STATEMENT;
+        return true;
+    }
+
+    bool scan_start_line_continuation(TSLexer *lexer) {
+      // Now see if we should start a line continuation
+      in_line_continuation = (lexer->lookahead == '&');
+      if (!in_line_continuation) {
+        return false;
+      }
+      // Consume the '&'
+      advance(lexer);
+      lexer->result_symbol = _LINE_CONTINUATION;
+      return true;
+    }
+
+    bool scan_end_line_continuation(TSLexer *lexer) {
+      if (!in_line_continuation) {
+        return false;
+      }
+      // Everything except comments ends a line continuation
+      if (lexer->lookahead == '!') {
+        return false;
+      }
+
+      in_line_continuation = false;
+
+      // Consume any leading line continuation markers
+      if (lexer->lookahead == '&') {
         advance(lexer);
+      }
+      lexer->result_symbol = _LINE_CONTINUATION;
+      return true;
+    }
 
-        // Allow whitespace after line continuation
-        while (lexer->lookahead == ' ') {
-          skip(lexer);
-        }
+    bool scan_string_literal(TSLexer *lexer) {
+      const char opening_quote = lexer->lookahead;
 
-        // Allow comments after line continuation
-        if (lexer->lookahead == '!') return true;
+      if (opening_quote != '"' && opening_quote != '\'') {
+        return false;
+      }
 
-        // Consume end of line characters, we allow '\n', '\r\n' and
-        // '\r' to cover unix, MSDOS and old style Macintosh
-        if (lexer->lookahead == '\r') {
-          lexer->advance(lexer, false);
+      advance(lexer);
+      lexer->result_symbol = STRING_LITERAL;
+
+      while (lexer->lookahead != '\n') {
```

Comment:
Reading the tree-sitter docs on external scanners I think we might want to tack on a ` && !lexer->eof(lexer)` clause to this to avoid a possible infinite loop edge case `lexer->lookahead` returns the NUL byte and the conditional would get stuck as true. I might try and trigger this case but it's probably a good safe guard either way.

Reply:
I couldn't get it to trip but still not certain if it's truly impossible to cause an inifinite loop

Reply:
Yes, I wondered if it might be wise to add that to all the loops

Reply:
The ones doing a positive assertion should be fine since the NUL byte would break the loop.